Tentatively remove runtime_registration_wrapper from cuda futures

msimberg opened the pull request:
Tentatively removing it because removing it should also fix the regression in #5366.

This is untested and I want to see what the different CUDA and HIP builders say about this change. According to @DarkDeepBlue things seem to be fine without the wrapper.

I think the wrapper is mostly unneeded, but I continuations are probably still problematic without the wrapper. When run on a non-HPX thread they would fall back to trying to get the default pool which is done through the runtime pointer.

One possible solution to this is to not have the thread-local runtime pointer be thread-local anymore. @hkaiser, @sithhell what was the original motivation for making it thread-local? The pointer itself is only ever read (except when initialized or reset during init and finalize), so I don't see any performance benefits of having it be thread-local. Is there some other reason for it?

hkaiser commented:
> One possible solution to this is to not have the thread-local runtime pointer be thread-local anymore. @hkaiser, @sithhell what was the original motivation for making it thread-local? The pointer itself is only ever read (except when initialized or reset during init and finalize), so I don't see any performance benefits of having it be thread-local. Is there some other reason for it?

From what I can see, the runtime_pointer could be made global without any issues nowadays. Same is probably true for some of the other thread-local variables we have.

hkaiser commented:
> Tentatively removing it because removing it should also fix the regression in #5366.

Not sure why this should fix the performance regression. The runtime_wrapper was in there forever.

> This is untested and I want to see what the different CUDA and HIP builders say about this change. According to @DarkDeepBlue things seem to be fine without the wrapper.

Things should be fine as long as the future's shared state does not suspend.

> I think the wrapper is mostly unneeded, but I continuations are probably still problematic without the wrapper. When run on a non-HPX thread they would fall back to trying to get the default pool which is done through the runtime pointer.

Yes, I agree.

msimberg commented:
> > Tentatively removing it because removing it should also fix the regression in #5366.
> 
> Not sure why this should fix the performance regression. The runtime_wrapper was in there forever.

The regression was introduced by changes to the wrapper. Removing it is of course the heavy-handed way of fixing the regression, but it should still fix the regression. And if we really don't need the wrapper then fixing the regression by removing it seems fair to me.

msimberg commented:
@hkaiser I've pushed some preliminary changes to make the runtime pointers non-thread local. The CUDA tests at least seem to be running fine with and without the distributed runtime.

One thing that annoyed me was that with the distributed runtime enabled the runtime pointer has to be reset early in `stop`/`stop_helper`, because if it's reset in the destructor it seems like at least some of the performance counter destruction/deinitialization/whatever is run (again?) in the `runtime_distributed` destructor and it eventually hits an assertion. For symmetry it'd be nice if the pointers could be reset in the `runtime`/`runtime_distributed` destructors (they're initialized in the constructors). Do you think that would be possible (I realize this is superficial, I can explain it in more detail if you'd like)?

hkaiser commented:
> One thing that annoyed me was that with the distributed runtime enabled the runtime pointer has to be reset early in `stop`/`stop_helper`, because if it's reset in the destructor it seems like at least some of the performance counter destruction/deinitialization/whatever is run (again?) in the `runtime_distributed` destructor and it eventually hits an assertion. For symmetry it'd be nice if the pointers could be reset in the `runtime`/`runtime_distributed` destructors (they're initialized in the constructors). Do you think that would be possible (I realize this is superficial, I can explain it in more detail if you'd like)?

Have you been able to resolve this? I can have a look otherwise.

msimberg commented:
> > One thing that annoyed me was that with the distributed runtime enabled the runtime pointer has to be reset early in `stop`/`stop_helper`, because if it's reset in the destructor it seems like at least some of the performance counter destruction/deinitialization/whatever is run (again?) in the `runtime_distributed` destructor and it eventually hits an assertion. For symmetry it'd be nice if the pointers could be reset in the `runtime`/`runtime_distributed` destructors (they're initialized in the constructors). Do you think that would be possible (I realize this is superficial, I can explain it in more detail if you'd like)?
> 
> Have you been able to resolve this? I can have a look otherwise.

I did not attempt to resolve it (meaning it's working as it is, but the pointer is not reset in the destructor). I'd definitely appreciate if you could have a look. Would you like a broken version to show the problem?

hkaiser commented:
> Would you like a broken version to show the problem?

That might simplify diagnosing the issue, much appreciated.

msimberg commented:
> > Would you like a broken version to show the problem?
> 
> That might simplify diagnosing the issue, much appreciated.

I've pushed another commit which moves the deinit to the destructors. Specifically, what happens is that `hpx::performance_counters::registry::remove_counter_type` checks if the (local) runtime pointer is still available, and proceeds if it is. It then fails further down on the stack. On master the runtime pointer (both of them) is already reset at this point. Since the distributed runtime destructor happens before the local runtime destructor the local runtime pointer is still there. It may very well be that the fix is as simple as checking for the distributed runtime pointer instead of the local one, but I can't judge if that's actually the correct thing to do.

msimberg commented:
@hkaiser have you possibly had a chance to look at this? If not, should we go ahead without moving the deinit for the moment (i.e. keep the same behaviour as currently on master)?

msimberg commented:
I've rebased and dropped the bad commit. If this comes back clean this should be ready to go. I've kept the breaking commit here: https://github.com/msimberg/hpx/commit/67c40f8a4f5dab1ba8bdb18bb225ad98747d65c4 in case someone wants to come back to this.

msimberg commented:
Reformatted. Looks like CI was happy with this except for the formatting. So this time it should really be ready to go.
